recipes/LibriSpeech/quantization/train_kmeans.py

```diff
@@ -0,0 +1,258 @@
+            feats = feats.flatten(end_dim=-2)  # [BN, H]
+            self.feats.append(feats)
+            self.curr_batch_size += len(feats)
+            if self.curr_batch_size < self.hparams.kmeans_batch_size:
+                # Leave average loss unchanged
+                loss = torch.tensor(self.avg_train_loss)
+                # Workaround to keep consistency with standard supervised training
+                loss.requires_grad_()
+                return loss
+            self.feats = torch.cat(self.feats)
+            feats = self.feats[: self.hparams.kmeans_batch_size]
+            self.feats = [self.feats[self.hparams.kmeans_batch_size :]]
+            self.curr_batch_size = len(self.feats[0])
+
+        if stage == sb.Stage.TRAIN:
+            # Retrieve current centroids
+            old_cluster_centers = self.hparams.quantizer.cluster_centers
+
+            # Partial fit on current batch
+            self.hparams.quantizer.partial_fit(feats)
+
+            # For K-means the training loss is the drift between current centroids and old centroids
+            # If close to 0, it means that the training has converged
+            curr_cluster_centers = self.hparams.quantizer.cluster_centers
+            loss = (curr_cluster_centers - old_cluster_centers).norm()
+
+            # Workaround to keep consistency with standard supervised training
+            loss.requires_grad_()
+            self.optimizer_step += 1
+            assert self.optimizer_step == self.modules.quantizer.n_steps, (
+                self.optimizer_step,
+                self.modules.quantizer.n_steps,
+            )
+        else:
+            # For K-means the validation/test loss is the inertia
+            # The lower the inertia, the better should be the clustering
+            # It is useful to monitor progress across epochs
+            # However, when saving checkpoints we always keep the last one (i.e. max_keys=["epoch"])
+            loss = self.hparams.quantizer.inertia(feats)
+
+        return loss
+
+    def on_stage_start(self, stage, epoch=None):
+        """Gets called at the beginning of each epoch."""
+        if stage == sb.Stage.TRAIN:
+            # NOTE: not included in intra-epoch checkpoints
+            self.feats = []
+            self.curr_batch_size = 0
+
+    def on_stage_end(self, stage, stage_loss, epoch=None):
+        """Gets called at the end of each epoch."""
+        # Compute/store important stats
+        current_epoch = self.hparams.epoch_counter.current
+        stage_stats = {"loss": stage_loss}
+
+        if stage == sb.Stage.TRAIN:
+            self.avg_train_loss = 0.0
+            self.train_stats = stage_stats
+            self.stats_meta = {"epoch": epoch, "steps": self.optimizer_step}
+            if if_main_process():
+                self.checkpointer.save_and_keep_only(
+                    meta={"loss": stage_stats["loss"], "epoch": epoch},
+                    max_keys=["epoch"],
```

Comment:
JSYK the checkpointer will always save the most recent checkpoints, the max_keys is not needed here, though its fine to have it to make things explicit.